<br><br><center><h1 style="font-size:4em;color:#2467C0">Week 1: Soccer Data Analysis</h1></center>
<br>
<table>
<col width="550">
<col width="450">
<tr>
<td><img src="https://images.pexels.com/photos/1667583/pexels-photo-1667583.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260" align="middle" style="width:550px;height:360px;"/></td>
<td>
Esta semana, vamos usar um dataset aberto do site popular<a href="https://www.kaggle.com"> Kaggle</a>. Este <a href="https://www.kaggle.com/hugomathien/soccer">conjunto de dados do Futebol Europeu</a> tem mais de 25 mil jogos e mais de 10 mil jogadores profissionais das temporadas entre 2008 e 2016. 
<br>
<br>
Embora não entraremos em detalhes para o nosso exemplo, o conjunto de dados ainda tem atributos sobre atualizações semanais do jogo, escalação de equipe e eventos de jogo detalhados.
<br>
<br>
O objetivo do notebook da análise é caminhar de ponta a ponta no processo de analise de um dataset e introduzir o que vamos cobrir neste curso. Nosso processo analítico inclui alguns passos para exploração e limpeza do dataset, alguns passos para predição da performance dos jogadores usando estatística básica e alguns passos para agrupar tipos semelhantes usando machine learning.
<br>
<br>
Vamos iniciar nossa Jornada em Python para Data Science!
</td>
</tr>
</table>

## Iniciando
<br> Para começar, vamos precisar:
<ol>
<li>Baixar os dados de: <a href="https://www.kaggle.com/hugomathien/soccer">https://www.kaggle.com/hugomathien/soccer</a></li>
<li>Extrair o arquivo .zip</li>
    <li>Mover o arquivo extraído 'database.sqlite' para sua pasta week 1r</li>
</ol>

## Importando Bibliotecas
<br> Vamos começar importando as bibliotecas Python que vamos usar ao longo deste análise. Estas bibliotecas incluem:
<ul>
<li><b>sqllite3</b> para interagir com o banco de dados relacional local; and</li>
<li><b>pandas</b> and <b>numpy</b> para ingestão e manipulação.</li>
<li><b>matplotlib</b> para visualização dos dados</li>
<li>specific methods from <b>sklearn</b> para Machine Learning 
<li><b>customplot</b>, que contém funções customizadas que nós escrevemos para este notebook</li>

</ul>

In [1]:
import sqlite3
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.preprocessing import scale
from customplot import *

#### Ingestão de dados

Agora, nós precisamos ler os dados usando os comandos abaixo. 

<b>Nota:</b> Tenha certeza que você rodou a célula acima (shift + enter) que importa o que vamos precisar, antes de rodar a célula abaixo.

<b>df</b> é uma variável apontada para um dataframe do pandas. Nós vamos aprender sobre na próxima semana.

In [2]:
# Crie sua conexão.
cnx = sqlite3.connect('database.sqlite')
df = pd.read_sql_query("SELECT * FROM Player_Attributes", cnx)

<h1 style="font-size:2em;color:#2467C0">Explorando os dados</h1>

Vamos começar nossa exploração dos dados gerando simples estatísticas dos dados. 
<br><br> 
Vamos olhar para as colunas de dados usando um atributo do pandas chamado "columns".

In [3]:
df.columns

Index(['id', 'player_fifa_api_id', 'player_api_id', 'date', 'overall_rating',
       'potential', 'preferred_foot', 'attacking_work_rate',
       'defensive_work_rate', 'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'dribbling', 'curve', 'free_kick_accuracy',
       'long_passing', 'ball_control', 'acceleration', 'sprint_speed',
       'agility', 'reactions', 'balance', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'aggression', 'interceptions', 'positioning',
       'vision', 'penalties', 'marking', 'standing_tackle', 'sliding_tackle',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes'],
      dtype='object')

Agora vamos mostrar estatísticas simples do nosso conjunto de dados. Você precisará rodar cada celula para ter certeza de visualizar as saídas.

In [4]:
df.describe().transpose()

count           mean            std     min        25%  \
id                  183978.0   91989.500000   53110.018250     1.0   45995.25   
player_fifa_api_id  183978.0  165671.524291   53851.094769     2.0  155798.00   
player_api_id       183978.0  135900.617324  136927.840510  2625.0   34763.00   
overall_rating      183142.0      68.600015       7.041139    33.0      64.00   
potential           183142.0      73.460353       6.592271    39.0      69.00   
crossing            183142.0      55.086883      17.242135     1.0      45.00   
finishing           183142.0      49.921078      19.038705     1.0      34.00   
heading_accuracy    183142.0      57.266023      16.488905     1.0      49.00   
short_passing       183142.0      62.429672      14.194068     3.0      57.00   
volleys             181265.0      49.468436      18.256618     1.0      35.00   
dribbling           183142.0      59.175154      17.744688     1.0      52.00   
curve               181265.0      52.965675      18.255788     2.0      41.00   
free_kick_accuracy  183142.0      49.380950      17.831746     1.0      36.00   
long_passing        183142.0      57.069880      14.394464     3.0      49.00   
ball_control        183142.0      63.388879      15.196671     5.0      58.00   
acceleration        183142.0      67.659357      12.983326    10.0      61.00   
sprint_speed        183142.0      68.051244      12.569721    12.0      62.00   
agility             181265.0      65.970910      12.954585    11.0      58.00   
reactions           183142.0      66.103706       9.155408    17.0      61.00   
balance             181265.0      65.189496      13.063188    12.0      58.00   
shot_power          183142.0      61.808427      16.135143     2.0      54.00   
jumping             181265.0      66.969045      11.006734    14.0      60.00   
stamina             183142.0      67.038544      13.165262    10.0      61.00   
strength            183142.0      67.424529      12.072280    10.0      60.00   
long_shots          183142.0      53.339431      18.367025     1.0      41.00   
aggression          183142.0      60.948046      16.089521     6.0      51.00   
interceptions       183142.0      52.009271      19.450133     1.0      34.00   
positioning         183142.0      55.786504      18.448292     2.0      45.00   
vision              181265.0      57.873550      15.144086     1.0      49.00   
penalties           183142.0      55.003986      15.546519     2.0      45.00   
marking             183142.0      46.772242      21.227667     1.0      25.00   
standing_tackle     183142.0      50.351257      21.483706     1.0      29.00   
sliding_tackle      181265.0      48.001462      21.598778     2.0      25.00   
gk_diving           183142.0      14.704393      16.865467     1.0       7.00   
gk_handling         183142.0      16.063612      15.867382     1.0       8.00   
gk_kicking          183142.0      20.998362      21.452980     1.0       8.00   
gk_positioning      183142.0      16.132154      16.099175     1.0       8.00   
gk_reflexes         183142.0      16.441439      17.198155     1.0       8.00   

                         50%        75%       max  
id                   91989.5  137983.75  183978.0  
player_fifa_api_id  183488.0  199848.00  234141.0  
player_api_id        77741.0  191080.00  750584.0  
overall_rating          69.0      73.00      94.0  
potential               74.0      78.00      97.0  
crossing                59.0      68.00      95.0  
finishing               53.0      65.00      97.0  
heading_accuracy        60.0      68.00      98.0  
short_passing           65.0      72.00      97.0  
volleys                 52.0      64.00      93.0  
dribbling               64.0      72.00      97.0  
curve                   56.0      67.00      94.0  
free_kick_accuracy      50.0      63.00      97.0  
long_passing            59.0      67.00      97.0  
ball_control            67.0      73.00      97.0  
acceleration            69.0   

<h1 style="font-size:2em;color:#2467C0">Limpeza dos dados: Lidando com Dados Faltantes</h1>
Dados reais nunca estão limpos. Nós precisamos garantir que nossa limpeza, seja convertendo ou nos livrando de dados nulos ou ausentes.<br>
A próxima célula de código nos mostra se alguma das 183978 linhas tem valores nulos em uma das 42 colunas.

In [5]:
# Existe alguma linha com valores Nulos?
df.isnull().any().any(), df.shape

(True, (183978, 42))

Agora vamos tentar encontrar quantos dados em cada coluna são nulos.

In [9]:
df.isnull().sum(axis=0)

id                        0
player_fifa_api_id        0
player_api_id             0
date                      0
overall_rating          836
potential               836
preferred_foot          836
attacking_work_rate    3230
defensive_work_rate     836
crossing                836
finishing               836
heading_accuracy        836
short_passing           836
volleys                2713
dribbling               836
curve                  2713
free_kick_accuracy      836
long_passing            836
ball_control            836
acceleration            836
sprint_speed            836
agility                2713
reactions               836
balance                2713
shot_power              836
jumping                2713
stamina                 836
strength                836
long_shots              836
aggression              836
interceptions           836
positioning             836
vision                 2713
penalties               836
marking                 836
standing_tackle     

## Ajustando valores nulos deletando-os

Nas próximas duas linhas, nos vamos largar os valores nulos percorrendo cada linha.

In [10]:
# Ajuste de valores nulos

# Pegue o valor inicial da quantidade de linhas antes de 
rows = df.shape[0]

# Largue as linhas com valores nulos
df = df.dropna()

Agora se nós checarmos os valores nulos e números de lihas, nós veremos que não existem valores nulos e o número de linhas diminuiu em conformidade com nossa ação anterior.

In [11]:
# Verifique se todos os nulos saíram do dataset?
print('Quantidade de linhas anterior: ', rows)
df.isnull().any().any(), df.shape

Quantidade de linhas anterior:  183978


(False, (180354, 42))

Para encontrar exatamente quantas linhas nós removemos, precisamos subtrair o número de linhas atual do número de linhas antes de remover os valores nulos.

In [12]:
# Quantas linhas com valores nulos foram removidas?

rows - df.shape[0]

3624

In [15]:
# Uma amostra do nosso dataset
df.head().transpose()

0                    1  \
id                                     1                    2   
player_fifa_api_id                218353               218353   
player_api_id                     505942               505942   
date                 2016-02-18 00:00:00  2015-11-19 00:00:00   
overall_rating                        67                   67   
potential                             71                   71   
preferred_foot                     right                right   
attacking_work_rate               medium               medium   
defensive_work_rate               medium               medium   
crossing                              49                   49   
finishing                             44                   44   
heading_accuracy                      71                   71   
short_passing                         61                   61   
volleys                               44                   44   
dribbling                             51                   51   
curve                                 45                   45   
free_kick_accuracy                    39                   39   
long_passing                          64                   64   
ball_control                          49                   49   
acceleration                          60                   60   
sprint_speed                          64                   64   
agility                               59                   59   
reactions                             47                   47   
balance                               65                   65   
shot_power                            55                   55   
jumping                               58                   58   
stamina                               54                   54   
strength                              76                   76   
long_shots                            35                   35   
aggression                            71                   71   
interceptions                         70                   70   
positioning                           45                   45   
vision                                54                   54   
penalties                             48                   48   
marking                               65                   65   
standing_tackle                       69                   69   
sliding_tackle                        69                   69   
gk_diving                              6                    6   
gk_handling                           11                   11   
gk_kicking                            10                   10   
gk_positioning                         8                    8   
gk_reflexes                            8                    8   

                                       2                    3  \
id                                     3                    4   
player_fifa_api_id                218353               218353   
player_api_id                     505942               505942   
date                 2015-09-21 00:00:00  2015-03-20 00:00:00   
overall_rating                        62                   61   
potential                             66                   65   
preferred_foot                     right                right   
attacking_work_rate               medium               medium   
defensive_work_rate               medium               medium   
crossing                              49                   48   
finishing                             44                   43   
heading_accuracy                      71                   70   
short_passing                         61                   60   
volleys                               44                   43   
dribbling                             51                   50   
curve                                 45                   44   
free_kick_accuracy                    39                   38   
long_passing                          64                   63   
ball_con

Nossa tabela de dados tem muitas linhas, como você pode ver. Nós podemos apenas olhar algumas linhas uma vez. Em vez de olhar as mesmas 10 linhas do topo todas as vezes, nós misturamos, então vamos ver diferentes exemplos de linhas no topo, de forma aleatória. Desta forma, nós garantimos que os dados não estão em uma ordem particular quando tentamos pegar uma amostra (como pegar as primeiras ou ultimas linhas) porque misturamos as linhas de forma aleatória.

In [16]:
# Misturamos as linhas do dataset então nós pegaremos exemplos distribuidos quando mostrar o topo das linhas.

df = df.reindex(np.random.permutation(df.index))

In [28]:
# Vamos verificar se mudou o head?
df.head().transpose()

78452                102330  \
id                                 78453               102331   
player_fifa_api_id                157006               172695   
player_api_id                      75216                39299   
date                 2012-02-22 00:00:00  2015-09-25 00:00:00   
overall_rating                        68                   66   
potential                             68                   66   
preferred_foot                     right                right   
attacking_work_rate               medium               medium   
defensive_work_rate               medium               medium   
crossing                              72                   36   
finishing                             61                   29   
heading_accuracy                      70                   55   
short_passing                         64                   54   
volleys                               52                   16   
dribbling                             68                   45   
curve                                 69                   24   
free_kick_accuracy                    51                   14   
long_passing                          65                   50   
ball_control                          65                   55   
acceleration                          64                   70   
sprint_speed                          70                   70   
agility                               67                   70   
reactions                             74                   55   
balance                               70                   70   
shot_power                            61                   56   
jumping                               90                   91   
stamina                               71                   70   
strength                              59                   75   
long_shots                            66                   11   
aggression                            54                   61   
interceptions                         47                   67   
positioning                           75                   20   
vision                                65                   29   
penalties                             62                   30   
marking                               30                   67   
standing_tackle                       40                   68   
sliding_tackle                        28                   68   
gk_diving                             10                    8   
gk_handling                           13                    8   
gk_kicking                            12                    7   
gk_positioning                        14                    9   
gk_reflexes                           10                   12   

                                  1772                 105466  \
id                                  1773               105467   
player_fifa_api_id                174665               198706   
player_api_id                      43317               198566   
date                 2014-09-18 00:00:00  2015-07-03 00:00:00   
overall_rating                        64                   73   
potential                             65                   81   
preferred_foot                     right                right   
attacking_work_rate                  low               medium   
defensive_work_rate               medium                  low   
crossing                              34                   63   
finishing                             30                   72   
heading_accuracy                      63                   44   
short_passing                         46                   79   
volleys                               41                   57   
dribbling                             46                   73   
curve                                 36                   65   
free_kick_accuracy                    23                   72   
long_passing                          39                   77   
bal

<h1 style="font-size:2em;color:#2467C0">Previsão: 'overall_rating' de um jogador</h1>
Agora que nosso passo de limpar o dataset esta razoavelmente completo e podemos confiar e entender mais os dados, nós começaremos mergulhando ainda mais no dataset. 

### Vamos dar uma olhada em algumas linhas.

Nós usamos a função head() para essa tarefa, iso nos dá todas as colunas de todas as linhas do top.

In [29]:
df.head(5)

id  player_fifa_api_id  player_api_id                 date  \
78452    78453              157006          75216  2012-02-22 00:00:00   
102330  102331              172695          39299  2015-09-25 00:00:00   
1772      1773              174665          43317  2014-09-18 00:00:00   
105466  105467              198706         198566  2015-07-03 00:00:00   
85405    85406              197813         210142  2015-09-21 00:00:00   

        overall_rating  potential preferred_foot attacking_work_rate  \
78452             68.0       68.0          right              medium   
102330            66.0       66.0          right              medium   
1772              64.0       65.0          right                 low   
105466            73.0       81.0          right              medium   
85405             74.0       76.0          right                high   

       defensive_work_rate  crossing  ...  vision  penalties  marking  \
78452               medium      72.0  ...    65.0       62.0     30.0   
102330              medium      36.0  ...    29.0       30.0     67.0   
1772                medium      34.0  ...    45.0       43.0     57.0   
105466                 low      63.0  ...    71.0       71.0     23.0   
85405               medium      63.0  ...    73.0       69.0     69.0   

        standing_tackle  sliding_tackle  gk_diving  gk_handling  gk_kicking  \
78452              40.0            28.0       10.0         13.0        12.0   
102330             68.0            68.0        8.0          8.0         7.0   
1772               73.0            53.0       11.0          9.0        10.0   
105466             24.0            25.0        5.0          7.0         5.0   
85405              73.0            65.0        8.0          8.0        14.0   

        gk_positioning  gk_reflexes  
78452             14.0         10.0  
102330             9.0         12.0  
1772              10.0         10.0  
105466            11.0         10.0  
85405             15.0         11.0  

[5 rows x 42 columns]

Na maior parte do tempo, nós apenas estamos interessados em plotar algumas colunas. Neste caso, nós podemos usar a opção de seleção de colunas do pandas como mostrado abaixo. Por favor ignore a primeira coluna na saída da célula abaixo. É o identificador exclusivo que age como index para os dados.<br><br>
<b>Nota:</b> A partir deste ponto, nós vamos começar nos referindo as colunas como 'features' na nossa descrição.

In [31]:
df[:10][['penalties', 'overall_rating']]

penalties  overall_rating
78452        62.0            68.0
102330       30.0            66.0
1772         43.0            64.0
105466       71.0            73.0
85405        69.0            74.0
97516        42.0            58.0
132537       80.0            81.0
55823        45.0            59.0
73131        67.0            71.0
143880       53.0            60.0

## Análise de correlação de Features 
Vamos checar se 'penalties' está correlacionado a 'overall_rating'. Nós usaremos uma operação de seleção semelhante, mas desta vez para todas as linhas e dentro da função de correlação.

# Eles estão correlacionados (usando o coeficiente de correlação de Pearson)?

In [32]:
df['overall_rating'].corr(df['penalties'])

0.39271510791118835

Vemos que o coeficiente de correlação de Pearson para essas duas colunas é de 0,39.<br><br>
Pearson vai de -1 a +1. Um valor de 0 diria que não há correlação, então não devemos nos preocupar em olhar para esse atributo. Um valor de 0,39 mostra alguma correlação, embora possa ser mais forte.<br><br>
Pelo menos, temos esses atributos que estão ligeiramente correlacionados. Isso nos dá esperança de que possamos construir um preditor significativo usando esses recursos correlacionados "fracamente".<br><br>
Agora, vamos criar uma lista das features que nós gostariamos de iterar para a mesma operação.

## Crie uma lista de potenciais features que você quer medir a correlação

In [33]:
potentialFeatures = ['acceleration', 'curve', 'free_kick_accuracy', 'ball_control', 'shot_power', 'stamina']

O for loop abaixo mostra na tela os coeficientes de correlação de 'overall_rating' de um jogador com cada feature que nós adicionamos a lista potencial acima.

In [44]:
# Verifique como as features são correlacionadas com overall_rating
for f in potentialFeatures:
    cor = df['overall_rating'].corr(df[f])
    print(f'{f}: {cor:.6f}')

acceleration: 0.243998
curve: 0.357566
free_kick_accuracy: 0.349800
ball_control: 0.443991
shot_power: 0.428053
stamina: 0.325606


## Quais features tem as maiores correlações com overall_rating?

Olhando para os valores mostrados na célular anterior, notamos que são duas features "ball_control" (0.44) e "shot_power" (0.43). Então estas duas features parecem ter maior correlação com overall_rating.


<h1 style="font-size:2em;color:#2467C0">Visualização de Dados:</h1>
Agora nos vamos começar a plotar os coeficientes de correlação para cada feature com overall_rating. Nós começamos selecionando as colunas e criando uma lista com com os coeficientes de correlação, chamado "correlation".

In [45]:
cols = ['potential',  'crossing', 'finishing', 'heading_accuracy',
       'short_passing', 'volleys', 'dribbling', 'curve', 'free_kick_accuracy',
       'long_passing', 'ball_control', 'acceleration', 'sprint_speed',
       'agility', 'reactions', 'balance', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'aggression', 'interceptions', 'positioning',
       'vision', 'penalties', 'marking', 'standing_tackle', 'sliding_tackle',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes']

In [46]:
# Crie uma lista contendo os coeficientes de correlação de Pearson entre overall_rating com cada coluna em cols
correlations = [ df['overall_rating'].corr(df[f]) for f in cols ]

In [47]:
len(cols), len(correlations)

(34, 34)

Com o comando acima, garantimos que o número de features selecionadas e os cálculos de correlação são os mesmos, neste caso, ambos tem 34 valores. As próximas duas células mostram algumas linhas de código que usa o funções do pandas para criar gráficos 2D desses valores de correlação e nome das colunas.

In [48]:
# Crie uma função para plotar um dataframe com as colunas em string e valores numericos

def plot_dataframe(df, y_label):  
    color='coral'
    fig = plt.gcf()
    fig.set_size_inches(20, 12)
    plt.ylabel(y_label)

    ax = df.correlation.plot(linewidth=3.3, color=color)
    ax.set_xticks(df.index)
    ax.set_xticklabels(df.attributes, rotation=75); #Notice the ; (remove it and see what happens !)
    plt.show()

In [49]:
# Crie um dataframe usando cols e correlations

df2 = pd.DataFrame({'attributes': cols, 'correlation': correlations}) 

In [ ]:
# Vamos plotar o dataframe acima usando a função que criamos
    
plot_dataframe(df2, 'Player\'s Overall Rating')

<h1 style="font-size:1.5em;color:#FB41C4">Analysis of Findings</h1>

Now it is time for you to analyze what we plotted. Suppose you have to predict a player's overall rating. Which 5 player attributes would you ask for?
<br><br>
<b>Hint:</b> Which are the five features with highest correlation coefficients?

<h1 style="font-size:2em;color:#2467C0">Clustering Players into Similar Groups</h1>

Until now, we used basic statistics and correlation coefficients to start forming an opinion, but can we do better? What if we took some features and start looking at each player using those features? Can we group similar players based on these features? Let's see how we can do this. 

<b>Note:</b> Generally, someone with domain knowledge needs to define which features. We could have also selected some of the features with highest correlation with overall_rating. However, it does not guarantee best outcome always as we are not sure if the top five features are independent. For example, if 4 of the 5 features depend on the remaining 1 feature, taking all 5 does not give new information.

## Select Features on Which to Group Players

In [ ]:
# Define the features you want to use for grouping players

select5features = ['gk_kicking', 'potential', 'marking', 'interceptions', 'standing_tackle']
select5features

In [ ]:
# Generate a new dataframe by selecting the features you just defined

df_select = df[select5features].copy(deep=True)

In [ ]:
df_select.head()

## Perform KMeans Clustering

Now we will use a machine learning method called KMeans to cluster the values (i.e., player features on `gk_kicking`, `potential`, `marking`, `interceptions`, and `standing_tackle`). We will ask for four clusters. We will talk about KMeans clustering and other machine learning tools in Python in Week 7 so we won't discuss these methods here.

In [ ]:
# Perform scaling on the dataframe containing the features

data = scale(df_select)

# Define number of clusters
noOfClusters = 4

# Train a model
model = KMeans(init='k-means++', n_clusters=noOfClusters, n_init=20).fit(data)

In [ ]:
print(90*'_')
print("\nCount of players in each cluster")
print(90*'_')

pd.value_counts(model.labels_, sort=False)

If you find that the below cell runs into an error, make sure you are running the latest version of `pandas` on your computer (try `pip install pandas` in your terminal). Save and restart your Jupyter session once you have updated `pandas`.

In [ ]:
# Create a composite dataframe for plotting
# ... Use custom function declared in customplot.py (which we imported at the beginning of this notebook)

P = pd_centers(featuresUsed=select5features, centers=model.cluster_centers_)
P

<h1 style="font-size:2em;color:#2467C0">Visualization of Clusters</h1>
We now have 4 clusters based on the features we selected, we can treat them as profiles for similar groups of players. We can visualize these profiles by plotting the centers for each cluster, i.e., the average values for each featuere within the cluster. We will use matplotlib for this visualization. We will learn more about matplotlib in Week 5. 

In [ ]:
# For plotting the graph inside the notebook itself, we use the following command

%matplotlib inline

In [ ]:
parallel_plot(P)

<h1 style="font-size:1.5em;color:#FB41C4">Analysis of Findings</h1>
### Can you identify the groups for each of the below?

<ul>
<li>Two groups are very similar except in `gk_kicking` - these players can coach each other on `gk_kicking`, where they differ.</li>
<li>Two groups are somewhat similar to each other except in potential.</li>
</ul>